packages/xchain-radix/src/client.ts

```diff
@@ -0,0 +1,852 @@
+import { bech32m } from 'bech32'
+import BIP32Factory, { BIP32Interface } from 'bip32'
+import { derivePath } from 'ed25519-hd-key'
+import * as ecc from 'tiny-secp256k1'
+// eslint-disable-next-line ordered-imports/ordered-imports
+import {
+  RadixChain,
+  XRD_DECIMAL,
+  assets,
+  bech32Lengths,
+  bech32Networks,
+  feesEstimationPublicKeys,
+  xrdRootDerivationPaths,
+} from './const'
+
+const xChainJsNetworkToRadixNetworkId = (network: Network): number => {
+  switch (network) {
+    case Network.Mainnet:
+    case Network.Stagenet:
+      return NetworkId.Mainnet
+    case Network.Testnet:
+      return NetworkId.Stokenet
+  }
+}
+
+type PartialTransactionPreviewResponse = {
+  receipt: {
+    status: 'Succeeded' | 'Failed' | 'Rejected'
+    fee_summary: {
+      execution_cost_units_consumed: number
+      finalization_cost_units_consumed: number
+      xrd_total_execution_cost: string
+      xrd_total_finalization_cost: string
+      xrd_total_royalty_cost: string
+      xrd_total_storage_cost: string
+      xrd_total_tipping_cost: string
+    }
+  }
+}
+
+/**
+ * The main client for the Radix network which is then wrapped by the {Client} adapting it to have
+ * a {BaseXChainClient} interface.
+ */
+export class RadixSpecificClient {
+  private innerNetwork: number
+  private innerGatewayClient: GatewayApiClient
+
+  constructor(networkId: number) {
+    this.innerNetwork = networkId
+    this.innerGatewayClient = RadixSpecificClient.createGatewayClient(networkId)
+  }
+
+  // #region Getters & Setters
+  public set networkId(networkId: number) {
+    this.innerNetwork = networkId
+    this.innerGatewayClient = RadixSpecificClient.createGatewayClient(networkId)
+  }
+
+  public get networkId(): number {
+    return this.innerNetwork
+  }
+
+  public get gatewayClient(): GatewayApiClient {
+    return this.innerGatewayClient
+  }
+  // #endregion
+
+  // #region Public Methods
+  public async currentEpoch(): Promise<number> {
+    return this.innerGatewayClient.status.getCurrent().then((status) => status.ledger_state.epoch)
+  }
+
+  public async currentStateVersion(): Promise<number> {
+    return this.innerGatewayClient.status.getCurrent().then((status) => status.ledger_state.state_version)
+  }
+
+  public async fetchBalances(address: string): Promise<Balance[]> {
+    const fungibleResources = await this.fetchFungibleResources(address)
+    const nonFungibleResources = await this.fetchNonFungibleResources(address)
+    const fungibleBalances = this.convertResourcesToBalances(fungibleResources)
+    const nonFungibleBalances = this.convertResourcesToBalances(nonFungibleResources)
+    return fungibleBalances.concat(nonFungibleBalances)
+  }
+
+  private convertResourcesToBalances(
+    resources: FungibleResourcesCollectionItem[] | NonFungibleResourcesCollectionItem[],
+  ): Balance[] {
+    const balances: Balance[] = []
+    // Iterate through nonFungibleResources
+    const decimalDiff = XRD_DECIMAL - 8
```

Comment:
Why is this decimalDiff needed? AssetAmount and BaseAmount can handle assets with non 8 decimals

Reply:
got it